zebra-chain/src/addresses/sapling.rs

```diff
@@ -1 +1,187 @@
 //! Sapling Shielded Payment Address types.
+
+use std::{fmt, io};
+
+use bech32::{self, FromBase32, ToBase32};
+
+#[cfg(test)]
+use proptest::{arbitrary::Arbitrary, array, prelude::*};
+
+use crate::{
+    keys::sapling,
+    serialization::{ReadZcashExt, SerializationError, ZcashDeserialize, ZcashSerialize},
+    Network,
+};
+
+/// Human-Readable Parts for input to bech32 encoding.
+mod human_readable_parts {
+    pub const MAINNET: &str = "zs";
+    pub const TESTNET: &str = "ztestsapling";
+}
+
+///
+#[derive(Clone, Copy, PartialEq)]
+pub struct SaplingShieldedAddress {
+    diversifier: sapling::Diversifier,
+    transmission_key: sapling::TransmissionKey,
+}
+
+impl fmt::Debug for SaplingShieldedAddress {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("SaplingShieldedAddress")
+            .field("diversifier", &self.diversifier)
+            .field("transmission_key", &self.transmission_key)
+            .finish()
+    }
+}
+
+impl SaplingShieldedAddress {
+    fn to_human_readable_address(&self, network: Network) -> Result<String, bech32::Error> {
```

Comment:
Should this be used internally to a `Display` impl? Is there a reason that it shouldn't be just a `Display` impl? (e.g., is it called elsewhere in some context where `Display` can't be used?)

The function takes a `network` parameter.  Should this be stored as part of the address structure itself, so that the address contains all the information required to render itself?

Reply:
Unfortunately the actual structure of Sapling addresses is technically independent of the network (unlike Sprout), so when they are serialized/deserialized on the wire the network is not involved at all. It's only when they are pretty-printed for human consumption that the network is thrown in basically to help the consuming software look at the right data elsewhere, I think @str4d told me once that the params for prod/test networks in Sapling are the same so this is why???

![image](https://user-images.githubusercontent.com/552961/78724035-8e226f80-78fa-11ea-9688-1676615a3452.png)
https://zips.z.cash/protocol/protocol.pdf#saplingpaymentaddrencoding

Reply:
And this holds for all the Sapling keys etc that define an encoding.

Reply:
😢 

Maybe it's possible to figure out some kind of workaround.

Reply:
> Unfortunately the actual structure of Sapling addresses is technically independent of the network (unlike Sprout)

It's true for Sprout as well. It only happens to be the case that Sprout addresses are serialized with a "prefix" such that when Base58Check-encoding them, you get the nice human-readable prefix chars. Those prefix chars are an implementation detail, and the Sprout address itself is network-independent.

Reply:
> I think @str4d told me once that the params for prod/test networks in Sapling are the same so this is why???

Even if the params were different, the addresses would still be technically network-independent. It would only be if e.g. testnet used a different pairing-friendly curve that the addresses would need to be altered.

Reply:
Right, the network is encoded into the serialization, which is not the same for Sapling addresses.

Reply:
The main design choice here is whether you are trying to represent the Sapling protocol structure, or the user-facing address. I think you're trying to represent the latter, in which case it makes sense to me to have a struct that encodes the network type somewhere (either as a type parameter or an internal field).

Reply:
It's annoying to have to have two structs for Sapling whereas for Sprout I only needed one. :(

Reply:
OK! After talking with @str4d, I was confused about how bech32 handles the human-readable parts on encode and decode, it should be fine to associate a network field in the struct just as we do for `SproutShieldedAddress`s.

Reply:
Actually, this only helps us when the address is bech32 encoded, but when it's not, it still doesn't include the network data.

Reply:
In practice, we do associate the use of an address with a particular network, so even though the "raw encoding" doesn't include that data, it's reasonable for the struct to do so. The raw encoding for Sapling is only used publicly as the input to / output of Bech32 encoding (we also use it in `zcashd` to move `PaymentAddress` structs across the FFI, but that's an internal implementation detail).

Reply:
OK so. We only encoded network for bech32. The actual raw encoding doesn't include it, for any Sapling types. For addresses, where we really only care about the bech32 encoding, we can just forego our normal `Zcash(De)Serialize` impls and just to 'to human readable'. 

What about the other key types? The `IncomingViewingKey`s, `FullViewingKey`s, and `SpendingKey`s also have this issue.

Reply:
We may want the 'human readable' equivalent of Zcash(De)Serialize instead for these.
None of those key types are inputs to transactions or messages.

Reply:
OK! We're basically going to copy what we've done in Sprout types by including the network field and implementing `Display` for all of them as the human-readable component.